doc: nrf: Adding documentation support for Monitor mode operation

VivekUppunda opened the pull request:
This change brings in documentation support for Monitor mode sniffer operation

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-13821/1/)
- Skipped

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-13821/1/)
- Skipped



*Note: This message is automatically posted and updated by the CI*

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-13821
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

sachinthegreen commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+   :local:
+   :depth: 2
+
+The nRF70 Series device supports Monitor mode, that is, listening for and receiving all IEEE 802.11 traffic on a given channel, not restricted to a certain SSID. With Monitor mode raw IEEE 802.11 packets can be received directly at the user application layer.
```
```suggestion
The nRF70 Series devices support monitor mode. In monitor mode the nRF70 series devices listen on a configured channel and receive all IEEE 802.11 traffic (i.e. belonging to all BSSIDs operating) on that channel. Traffic belonging to all BSSIDs operating on that channel is received and passed . The captured traffic is then piped to the application layer using raw sockets, after prepending it with some metadata.
```

VivekUppunda replied:
This was changed as per earlier discussion to keep it concise

sachinthegreen commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+The nRF70 device has to be configured in monitor mode in order to operate as an 802.11 wireless packet sniffer.
+
+In Monitor mode, raw 802.11 packets received at the nRF70 Series device may be presented directly at the host application layer. When forwarding the received IEEE 802.11 packets to the user application, the Wi-Fi host driver will append a raw RX header to each received packet and deliver it up to the application layer using the raw socket API.
+The raw RX header holds the receive parameters for the received packet, such as data rate, receive signal strength, frequency of reception, and packet type (whether it is Legacy, HT, VHT, or HE).
```
Not sure if the above lines are necessary (assuming that the above suggestion is acceptable).

VivekUppunda replied:
This is as per previous review comments

sachinthegreen commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+
+.. _ug_nrf70_developing_enabling_802.11_monitor_mode:
+
+Enabling Monitor mode
```
```suggestion
Enabling monitor mode
```

VivekUppunda replied:
Following documentation convention that all modes are to begin with capital lettering

sachinthegreen commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+Enabling Monitor mode
+*********************
+
+To be able to use the Monitor mode in your application, you need to enable the :kconfig:option:`CONFIG_NRF700X_RAW_DATA_RX` Kconfig option in the project configuration.
```
```suggestion
To be able to use the monitor mode in your application, you need to enable the :kconfig:option:`CONFIG_NRF700X_RAW_DATA_RX` Kconfig option in the project configuration.
```

VivekUppunda replied:
Using the documentation convention that all modes will start with a capital letter.

sachinthegreen commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+Monitor mode operation
+**********************
+
+To enable the Monitor mode in the nRF Wi-Fi driver at runtime, use the ``NET_REQUEST_WIFI_MODE`` network management runtime API.
```
```suggestion
Monitor mode can be enabled or disabled on the nRF70 device at runtime, using the ``NET_REQUEST_WIFI_MODE`` network management runtime API.
```

sachinthegreen commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+**********************
+
+To enable the Monitor mode in the nRF Wi-Fi driver at runtime, use the ``NET_REQUEST_WIFI_MODE`` network management runtime API.
+This runtime API can be used to disable the Monitor mode in the driver, when packet sniffer operation is not required anymore in the user application.
```
This statement can be removed if the above suggestion is acceptable.

VivekUppunda replied:
This was broken into 2 sentences as I followed similar wording for the TX injection documentation.

sachinthegreen commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+To enable the Monitor mode in the nRF Wi-Fi driver at runtime, use the ``NET_REQUEST_WIFI_MODE`` network management runtime API.
+This runtime API can be used to disable the Monitor mode in the driver, when packet sniffer operation is not required anymore in the user application.
+
+When the nRF70 device is configured to operate in Monitor mode, simultaneous connected Station operation is not permitted (the nRF70 Series device operates exclusively as a Wi-Fi sniffer).
```
```suggestion
Monitor mode on the nRF70 device cannot operate concurrently with other modes such as station or softAP.
```

sachinthegreen commented on doc/nrf/protocols/wifi/index.rst:
```diff
@@ -37,3 +37,4 @@ If you want to go through an online training course to familiarize yourself with
    debugging
    raw_tx_operation
    sap
+   sniffer_rx_operation
```
Club this with `raw_tx_operation` (so that they appear one below the other on the page) and move both of them below `sap`

sachinthegreen commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+
+When the nRF70 device is configured to operate in Monitor mode, simultaneous connected Station operation is not permitted (the nRF70 Series device operates exclusively as a Wi-Fi sniffer).
+
+The user needs to configure the operating Wi-Fi channel on which the nRF70 device will monitor and receive 802.11 packets.
```
It is mandatory to configure an operating channel on the nRF70 device before monitor mode can be enabled on it. This can done using ``NET_REQUEST_WIFI_CHANNEL`` runtime NET-management API.

ioannisg replied:
The API is discussed below. 
"mandatory" is probably implied by "needs to"

VivekUppunda replied:
Yes, that is correct.

VivekUppunda commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+﻿.. _ug_nrf70_developing_raw_ieee_80211_packet_reception:
+
+Raw IEEE 802.11 packet reception using Monitor mode
```
This is the comment to leave all modes to begin with capital letter. I am following the same convention as before

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+   :local:
+   :depth: 2
+
+The nRF70 Series device supports Monitor mode, that is, listening for and receiving all IEEE 802.11 traffic on a given channel, not restricted to a certain SSID. With Monitor mode raw IEEE 802.11 packets can be received directly at the user application layer.
```
Please ensure "Monitor mode" is used throughout.

Also, Sachin's suggestion can be further condensed to the following:

```suggestion
The nRF70 Series devices support Monitor mode.
In this mode, the device listens on a configured channel and receive all IEEE 802.11 traffic, including traffic from all BSSIDs operating on that channel.
The device then prepends the captured traffic with metadata before passing it to the application layer using raw sockets.
```

ioannisg replied:
I think most of this is repeated below (e.g. the thing about raw sockets, packet meta-data, etc.). Therefore, I envisioned a simpler and more high-level description here.

VivekUppunda replied:
Yes, earlier it was a more elaborate description at the top. Hence, as per discussion, it was changed.

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+   :depth: 2
+
+The nRF70 Series device supports Monitor mode, that is, listening for and receiving all IEEE 802.11 traffic on a given channel, not restricted to a certain SSID. With Monitor mode raw IEEE 802.11 packets can be received directly at the user application layer.
+The nRF70 device has to be configured in monitor mode in order to operate as an 802.11 wireless packet sniffer.
```
Please ensure "Monitor mode" is used throughout.

```suggestion
You must configure the nRF70 device in Monitor mode to operate it as an 802.11 wireless packet sniffer.".
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+The nRF70 device has to be configured in monitor mode in order to operate as an 802.11 wireless packet sniffer.
+
+In Monitor mode, raw 802.11 packets received at the nRF70 Series device may be presented directly at the host application layer. When forwarding the received IEEE 802.11 packets to the user application, the Wi-Fi host driver will append a raw RX header to each received packet and deliver it up to the application layer using the raw socket API.
+The raw RX header holds the receive parameters for the received packet, such as data rate, receive signal strength, frequency of reception, and packet type (whether it is Legacy, HT, VHT, or HE).
```
```suggestion
The raw RX header contains parameters of the received packet, including data rate, signal strength, reception frequency, and packet type, and packet type (whether it is Legacy, HT, VHT, or HE).
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+The nRF70 Series device supports Monitor mode, that is, listening for and receiving all IEEE 802.11 traffic on a given channel, not restricted to a certain SSID. With Monitor mode raw IEEE 802.11 packets can be received directly at the user application layer.
+The nRF70 device has to be configured in monitor mode in order to operate as an 802.11 wireless packet sniffer.
+
+In Monitor mode, raw 802.11 packets received at the nRF70 Series device may be presented directly at the host application layer. When forwarding the received IEEE 802.11 packets to the user application, the Wi-Fi host driver will append a raw RX header to each received packet and deliver it up to the application layer using the raw socket API.
```
```suggestion
In Monitor mode, raw 802.11 packets received by the nRF70 Series device may be presented directly at the host application layer.
When forwarding the received IEEE 802.11 packets to the user application, the Wi-Fi host driver will append a raw RX header to each received packet and deliver it up to the application layer using the raw socket API.
```

VivekUppunda replied:
Removed Line 13 as it seemed like a repetition

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+**********************
+
+Monitor mode can be enabled or disabled on the nRF70 Series device at runtime, using the ``NET_REQUEST_WIFI_MODE`` network management runtime API.
+This runtime API can be used to disable the Monitor mode in the driver, when packet sniffer operation is not required anymore in the user application.
```
```suggestion
This runtime API can be used to disable the Monitor mode in the driver, when the packet sniffer operation is no longer required in the user application.
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+Monitor mode receive packet metadata
+************************************
+
+The IEEE 802.11 packet that is captured by the nRF70 Series device will prepend a proprietary header ``raw_rx_pkt_header`` which contains the following information about the received 802.11 packet.
```
```suggestion
The IEEE 802.11 packet that is captured by the nRF70 Series device will prepend a proprietary header ``raw_rx_pkt_header`` which contains the following information about the received 802.11 packet:
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+Monitor mode receive operation
+******************************
+
+An IEEE 802.11 packet captured by the nRF70 Series device in Monitor mode will be sent up to the Wi-Fi driver which will append a radio information header to the received packet and present the resulting set to the user application.
```
```suggestion
An IEEE 802.11 packet captured by the nRF70 Series device in Monitor mode will be sent up to the Wi-Fi driver, which will append a radio information header to the received packet and present the resulting set to the user application.
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+     - Parameter to enable/disable TX injection mode.
+       ``1`` enables TX injection and ``0`` disables TX injection.
+
+.. note::
```
I suggest moving this note to after line 116 - 

It also depends on, what is the order of steps to be able to enable TX injection mode?
Is it: 
1. Enable (kconfig option)
2. Invoke (API)  

And both steps are required? If so, I would describe it in that order (or which ever order is necessary), and not present it as a note, rather an instruction in this section.

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,136 @@
+The following figure illustrates the packet structure and IEEE 802.11 packet sniffer operational flow:
+
+.. figure:: images/nrf7000_packet_sniffer_operation.png
+   :alt: IEEE 802.11 packet sniffer operational flow
```
```suggestion
   :alt: IEEE 802.11 packet sniffer operational flow
   
   IEEE 802.11 packet sniffer operational flow
```

ioannisg reviewed (approved):
Looks good to me

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,138 @@
+The nRF70 Series devices support Monitor mode.
+You must configure the nRF70 device in Monitor mode to operate it as an 802.11 wireless packet sniffer.
+
+In this mode, the device listens on a configured channel and receive all IEEE 802.11 traffic, including traffic from all BSSIDs operating on that channel.
```
```suggestion
In this mode, the device listens on a configured channel and receives all IEEE 802.11 traffic, including traffic from all BSSIDs operating on that channel.
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,138 @@
+
+Monitor mode on the nRF70 device cannot operate concurrently with other modes such as Station mode or SoftAP mode.
+
+The user needs to configure the operating Wi-Fi channel on which the nRF70 device will monitor and receive 802.11 packets.
```
To be more in line with the tone of our doc:
```suggestion
You must configure the operating Wi-Fi channel on which the nRF70 device will monitor and receive 802.11 packets.
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,138 @@
+To set the desired channel for 802.11 packet reception, use the ``NET_REQUEST_WIFI_CHANNEL`` runtime NET-management API.
+The channel configuration runtime NET-management API can be used to set the channel after setting the nRF70 Series device in Monitor mode.
+
+See the :ref:`wifi_shell_sample` sample for more information on configuring mode and channel settings for raw packet reception through shell commands.
```
```suggestion
See the :ref:`wifi_shell_sample` sample for more information on configuring the mode and channel settings for raw packet reception through shell commands.
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,138 @@
+
+See the :ref:`wifi_shell_sample` sample for more information on configuring mode and channel settings for raw packet reception through shell commands.
+
+The following table lists the shell commands and NET-management APIs that are used to enable Monitor and TX injection mode and select the desired Wi-Fi channel:
```
```suggestion
The following table lists the shell commands and NET-management APIs that are used to enable Monitor and TX injection mode and select the desired Wi-Fi channel:
```
```suggestion
The following table lists the shell commands and NET-management APIs that are used to enable Monitor and TX injection mode, and to select the desired Wi-Fi channel:
```

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,138 @@
+The user needs to configure the operating Wi-Fi channel on which the nRF70 device will monitor and receive 802.11 packets.
+When the device operates in Monitor mode, all 802.11 packets received on the configured channel will be sent up the stack by the nRF70 Series device and the Wi-Fi driver.
+
+To set the desired channel for 802.11 packet reception, use the ``NET_REQUEST_WIFI_CHANNEL`` runtime NET-management API.
```
Should it be: Network management API?
I couldn't find NET-management used in other docs; if it is used elsewhere/is official name of the API then please leave as is.

kittydepa commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,138 @@
+
+You must enable the :kconfig:option:`CONFIG_NRF700X_RAW_DATA_TX` Kconfig option for TX injection mode of operation.
+
+TX injection mode can be enabled by invoking the API ``net_eth_txinjection_mode``.
```
```suggestion
TX injection mode can be also be enabled by invoking the API ``net_eth_txinjection_mode``.
```

or if both are required:

```suggestion
TX injection mode can be also needs to be enabled by invoking the API ``net_eth_txinjection_mode``.
```
?

VivekUppunda replied:
i have modified as below as one change is to enable the feature in code and the other is to enable the feature during runtime

You must enable the :kconfig:option:`CONFIG_NRF700X_RAW_DATA_TX` Kconfig option for TX injection mode of operation to be enabled in the build.
 
TX injection mode can be enabled during runtime by invoking the API ``net_eth_txinjection_mode``.

sachinthegreen reviewed (changes requested):
I think an entry for this would be needed in https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/releases_and_maturity/software_maturity.html#wi-fi-feature-support

I will check the same. I only see one entry. I will figure out how to add to sml-table and update the same

richabp commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,137 @@
+     - Signal strength of the received packet.
+   * - rate_flags
+     - unsigned char
+     - Provides data on whether the received 802.11 packet is a legacy, HT, VHT or HE packet.
```
```suggestion
     - Provides data on whether the received 802.11 packet is a legacy, HT, VHT, or HE packet.
```

richabp commented on doc/nrf/protocols/wifi/sniffer_rx_operation.rst:
```diff
@@ -0,0 +1,137 @@
+     - Provides the data rate at which the packet was received.
+       It can be a ``legacy rate`` or an ``MCS rate`` based on the ``rate_flags`` value.
+
+You can refer to the relevant structure at:
+:file:`nrfxlib/nrf_wifi/fw_if/umac_if/inc/default/fmac_structs.h` - for the raw RX header
```
Since there is only one entry so it would be better to present it in a single line rather than in list form..
```suggestion
You can refer to the relevant structure at :file:`nrfxlib/nrf_wifi/fw_if/umac_if/inc/default/fmac_structs.h` for the raw RX header.
```

richabp commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -1018,6 +1020,7 @@ Documentation
   * :ref:`ug_lte` user guide under :ref:`protocols`.
   * Gazell and NFC sections in the :ref:`app_power_opt_recommendations` user guide.
   * :ref:`ug_nrf70_stack partitioning` page in the :ref:`ug_nrf70_developing` user guide.
+  * A page on :ref:`ug_nrf70_developing_raw_ieee_80211_packet_reception` in the :ref:`ug_wifi` page.
```
```suggestion
  * A page on :ref:`ug_nrf70_developing_raw_ieee_80211_packet_reception` in the :ref:`ug_wifi` user guide.
```

VivekUppunda commented:
> I think an entry for this would be needed in https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/releases_and_maturity/software_maturity.html#wi-fi-feature-support
> 
> I will check the same. I only see one entry. I will figure out how to add to sml-table and update the same

As per current testing and discussion, i have added the 

> I think an entry for this would be needed in https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/releases_and_maturity/software_maturity.html#wi-fi-feature-support
> 
> I will check the same. I only see one entry. I will figure out how to add to sml-table and update the same

As per current testing and discussion, i have added BOARD_NRF7002DK_NRF5340_CPUAPP to the software_maturity_features.yaml file

Thank you
Vivek
